packages/core/contracts/financial-templates/optimistic-distributor/OptimisticDistributor.sol

```diff
@@ -0,0 +1,420 @@
+
+/**
+ * @title  OptimisticDistributor contract.
+ * @notice Allows sponsors to distribute rewards through MerkleDistributor contract secured by UMA Optimistic Oracle.
+ */
+contract OptimisticDistributor is Lockable, MultiCaller {
+    using SafeERC20 for IERC20;
+
+    /********************************************
+     *  OPTIMISTIC DISTRIBUTOR DATA STRUCTURES  *
+     ********************************************/
+
+    // Represents reward posted by a sponsor.
+    struct Reward {
+        address sponsor;
+        IERC20 rewardToken;
+        uint256 maximumRewardAmount;
+        uint256 earliestProposalTimestamp;
+        uint256 optimisticOracleProposerBond;
+        uint256 optimisticOracleLivenessTime;
+        bytes32 priceIdentifier;
+        bytes customAncillaryData;
+    }
+
+    // Represents proposed rewards distribution.
+    struct Proposal {
+        uint256 rewardIndex;
+        uint256 timestamp;
+        bytes32 merkleRoot;
+        string ipfsHash;
+    }
+
+    /********************************************
+     *      INTERNAL VARIABLES AND STORAGE      *
+     ********************************************/
+
+    // Reserve for bytes appended to ancillary data (e.g. OracleSpoke) when resolving price from non-mainnet chains.
+    // This also covers appending proposalIndex by this contract.
+    uint256 public constant ANCILLARY_BYTES_RESERVE = 512;
+
+    // Restrict Optimistic Oracle liveness to between 10 minutes and 100 years.
+    uint256 public constant MINIMUM_LIVENESS = 10 minutes;
+    uint256 public constant MAXIMUM_LIVENESS = 5200 weeks;
+
+    // Final fee can be synced and stored in the contract.
+    uint256 public finalFee;
+
+    // Index of next created reward or proposal.
+    uint256 public nextCreatedReward;
+    uint256 public nextCreatedProposal;
+
+    // Rewards and proposals are mapped to their indices.
+    mapping(uint256 => Reward) public rewards;
+    mapping(uint256 => Proposal) public proposals;
+
+    // Immutable variables provided at deployment.
+    FinderInterface public immutable finder;
+    IERC20 public bondToken; // This cannot be declared immutable as bondToken needs to be checked against whitelist.
+
+    // Merkle Distributor can be set only once.
+    MerkleDistributor public merkleDistributor;
+
+    // Interface parameters that can be synced and stored in the contract.
+    StoreInterface public store;
+    OptimisticOracleInterface public optimisticOracle;
+
+    /********************************************
+     *                  EVENTS                  *
+     ********************************************/
+
+    event RewardCreated(
+        address indexed sponsor,
+        IERC20 rewardToken,
+        uint256 indexed rewardIndex,
+        uint256 maximumRewardAmount,
+        uint256 earliestProposalTimestamp,
+        uint256 optimisticOracleProposerBond,
+        uint256 optimisticOracleLivenessTime,
+        bytes32 priceIdentifier,
```

Comment:
we can index up to 3. no point in not using all of them :)
```suggestion
        bytes32 indexed priceIdentifier,
```

Reply:
fixed in 1766ed802ff29f1371896ba5d54765ce2fd35e04